### ISSUE TITLE
segmentation fault in FromIterator impl

### ISSUE THREAD

**Issue by MaxOhn** [ibraheemdev/papaya#20]
I usually work on Windows which apparently isn't affected but as soon as I ran on Ubuntu or WSL I encountered a segfault.

After some quick checks I pinned it down to `papaya::HashMap`'s `FromIterator` implementation, specifically the line https://github.com/ibraheemdev/papaya/blob/ca9cf5e9cc88f6ac7425bc1235a7f62570d8afe0/src/map.rs#L1090

I struggled to create an MRE; the best I could come up with is

```rs
fn main() {
    // basically just a source for random values; should be about 500kb+ in size
    let data = std::fs::read("/some/decently/big/file.txt").unwrap();
    let mut map = std::collections::HashMap::new();

    for buf in data.chunks(5) {
        let n = buf.iter().fold(1, |prod, byte| (*byte as u64) * prod); // producing random values
        map.insert(n, ());
    }

    // segfault if the file was long enough to put ~7000+ entries into the map
    dbg!(map.len());

    let _ = map
        .into_iter()
        .collect::<papaya::HashMap<_, _, IntHasher>>(); // <- segfaults here
}

#[derive(Default)]
struct IntHasher;

struct IntHash(Option<u64>);

impl std::hash::Hasher for IntHash {
    fn write(&mut self, _: &[u8]) {
        unimplemented!()
    }

    fn write_u64(&mut self, n: u64) {
        self.0 = Some(n);
    }

    fn finish(&self) -> u64 {
        self.0.unwrap()
    }
}

impl std::hash::BuildHasher for IntHasher {
    type Hasher = IntHash;

    fn build_hasher(&self) -> Self::Hasher {
        IntHash(None)
    }
}
```

Output:

```sh
$ cargo run 
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
     Running `/target/debug/testy`
[src/main.rs:12:5] map.len() = 6588
zsh: segmentation fault  cargo run
```

Ran on cargo version `cargo 1.80.0 (376290515 2024-07-16)` in WSL with `papaya = "0.1.2"`.

**Comment by MaxOhn** [comment]
As a quick workaround, I copy-pasted the `FromIterator` implementation but instead of the `unsafe` line I create the guard the regular way.

https://github.com/MaxOhn/Bathbot/commit/7d39e458fba8aeda1b6be4d89e0ba80d12f042ec

**Comment by ibraheemdev** [comment]
Hmm this is tricky. You're right that the `seize::unprotected` is incorrect because `insert` returns a reference to the previous value if it was replaced, so we need a real guard there. We could probably add a `raw_insert` that doesn't return a reference, but using a guard will at least fix this bug.

**Comment by ibraheemdev** [comment]
Released 0.1.3 with the fix. Thanks for the report!
